### ISSUE TITLE
PWC-Net not to converge

### ISSUE THREAD

**Issue by Nemophilist98** [875159954/tartanvo_train_implementation#2]
Hello, I have run your code, and the result is basically the same as the problem you mentioned before [#issue](https://github.com/castacks/tartanvo/issues/21#issue-2453976943).  When I trained PWC-Net alone, I still failed to converge, and I found that the get_loss  was used. Would it be better if it was replaced with get_loss_w_mask?   Have you solved it now?

Looking forward to your reply.   Thank you!

**Comment by Zilong-L** [comment]
Sorry for the late reply. After I finished my project proposal in September, I relaxed and didn’t do much work.
I encountered repeated failures during the second stage training when trying to minimize the loss, which is why I decided not to continue with it. So I decided to start from scratch and work on my own approach, based on my efforts to reproduce TartanVO.
I do have some experience that I can share. I was able to achieve actual results with an end-to-end model. It is absolutely feasible to predict the camera pose directly from features obtained by images, and I fully believe that TartanVO can be successfully trained.

As for the loss replacement you mentioned, I didn’t come across this loss function in TartanVO, so you might be referring to something in DytanVO or the PWC-Net? I am not sure what you are talking about. If you have an idea, feel free to fork it and try replacing it.

Currently, I’m focused on my own end to end model, so I haven’t been able to continue with this repo. If I finish my current work and have time, I might revisit the TartanVO training code and give it another try.

**Comment by Nemophilist98** [comment]
Thank you for your reply. Meanwhile, I'm really grateful for your contributions. After that, I didn't continue with this work anymore. Recently, this week, I started to reproduce this work again. Here are the training results of mine today. As you can see, the training loss is gradually reaching the expected level. I used the PWC-Net network and set lambda to 0.05. Although this optical flow network doesn't seem to be able to converge well, I just ignored it and directly proceeded to the second stage of training. Most importantly, when calculating the translation loss in your code, the ground truth wasn't normalized. And this is my improvement:

```
def linear_norm_trans_loss(self, output, motion, mask=None):
        output_trans = output[:, :3]
        output_rot = output[:, 3:]
        trans_norm = torch.norm(output_trans, dim=1).view(-1, 1)
        output_norm = output_trans / trans_norm

        target_trans = motion[:, :3]
        target_trans_norm = torch.norm(target_trans, dim=1).view(-1, 1)
        target_norm = target_trans / target_trans_norm

        if mask is None:
            trans_loss = self.criterion(output_norm, target_norm)
            rot_loss = self.criterion(output_rot, motion[:, 3:])
        else:
            trans_loss = self.criterion(output_norm[mask,:], target_norm[mask,:])
            rot_loss = self.criterion(output_rot[mask,:], motion[mask, 3:])

        loss = (rot_loss + trans_loss)/2.0

        return loss, trans_loss.item() , rot_loss.item()
```
        
![1732091584285](https://github.com/user-attachments/assets/f1475c00-57c9-4049-990e-be9296b5c59c)
